libsql-sqlite3/src/vectorIndex.c

```diff
@@ -459,4 +462,498 @@ int parseVectorIdxParams(Parse *pParse, VectorIdxParams *pParams, int type, int
   return SQLITE_OK;
 }
 
+/**************************************************************************
+** Vector index cursor implementation
+****************************************************************************/
+
+/*
+** A VectorIdxCursor is a special cursor to perform vector index lookups.
+ */
+struct VectorIdxCursor {
+  sqlite3 *db;            /* Database connection */
+  DiskAnnIndex *pIndex;   /* DiskANN index */
+};
+
+
+void skipSpaces(const char **pzStr) {
+  while( **pzStr != '\0' && sqlite3Isspace(**pzStr) ){
+    (*pzStr)++;
+  }
+}
+/**
+** Parses a type string such as `FLOAT32(3)` and set number of dimensions and bits
+**
+** Returns  0 if succeed and set correct values in both pDims and pType pointers
+** Returns -1 if the type string is not a valid vector type for index and set pErrMsg to static string with error description in this case
+**/
+int vectorIdxParseColumnType(const char *zType, int *pType, int *pDims, const char **pErrMsg){
+  assert( zType != NULL );
+
+  int dimensions = 0;
+  int i;
+  skipSpaces(&zType);
+  for(i = 0; i < ArraySize(VECTOR_COLUMN_TYPES); i++){
+    const char* name = VECTOR_COLUMN_TYPES[i].zName;
+    const char* zTypePtr = zType + strlen(name);
+    if( sqlite3_strnicmp(zType, name, strlen(name)) != 0 ){
+      continue;
+    }
+    skipSpaces(&zTypePtr);
+    if( *zTypePtr != '(' ) {
+      break;
+    }
+    zTypePtr++;
+    skipSpaces(&zTypePtr);
+
+    while( *zTypePtr != '\0' && *zTypePtr != ')' && !sqlite3Isspace(*zTypePtr) ){
+      if( !sqlite3Isdigit(*zTypePtr) ){
+        *pErrMsg = "non digit symbol in vector column parameter";
+        return -1;
+      }
+      dimensions = dimensions*10 + (*zTypePtr - '0');
+      if( dimensions > MAX_VECTOR_SZ ) {
+        *pErrMsg = "max vector dimension exceeded";
+        return -1;
+      }
+      zTypePtr++;
+    }
+    skipSpaces(&zTypePtr);
+    if( *zTypePtr != ')' ){
+      *pErrMsg = "missed closing brace for vector column type";
+      return -1;
+    }
+    zTypePtr++;
+    skipSpaces(&zTypePtr);
+
+    if( *zTypePtr != '\0' ) {
+      *pErrMsg = "extra data after dimension parameter for vector column type";
+      return -1;
+    }
+
+    if( dimensions <= 0 ){
+      *pErrMsg = "vector column must have non-zero dimension for index";
+      return -1;
+    }
+
+    *pDims = dimensions;
+    if( VECTOR_COLUMN_TYPES[i].nBits == 32 ) {
+      *pType = VECTOR_TYPE_FLOAT32;
+    } else if( VECTOR_COLUMN_TYPES[i].nBits == 64 ) {
+      *pType = VECTOR_TYPE_FLOAT64;
+    } else {
+      *pErrMsg = "unsupported vector type";
+      return -1;
+    }
+    return 0;
+  }
+  *pErrMsg = "unexpected vector column type";
+  return -1;
+}
+
+int initVectorIndexMetaTable(sqlite3* db) {
+  static const char *zSql = "CREATE TABLE IF NOT EXISTS " VECTOR_INDEX_GLOBAL_META_TABLE " ( name TEXT, metadata BLOB );";
```

Comment:
Should we make name the primary key and add `WITHOUT ROWID`?

Reply:
Yes, thanks! New schema highlighted a bug: SQLite can reparse schema (for several reasons, for example validation that after `ALTER TABLE` schema still valid) - and we don't want to insert anything in this case